### ISSUE TITLE
Apple Silicon Universal Build

### ISSUE THREAD

**Issue by semihalev** [AdguardTeam/AdGuardForSafari#493]
Hey Team,

Firstly thanks for the amazing extension. I just checked, there is no universal build for latest M1 chipset. Are there any plan about this?

**Comment by TheHasagi** [comment]
@semihalev 

Hello, unfortunately, Electron is not supporting it yet. The best way is to launch AdGuard for Safari with Rosetta on the M1 platform.

**Comment by F3000** [comment]
Look here : 

Version 11 is supporting M1 Macs out of the box 

https://www.electronjs.org/blog/electron-11-0

**Comment by TheHasagi** [comment]
Supporting M1, but not the universal builds.

**Comment by semihalev** [comment]
Can you build separately like arm64 and x64? I seen Electron already supported arm64 on darwin.

https://github.com/electron/electron/releases/tag/v11.1.1

**Comment by ameshkov** [comment]
We could try, but we won't be able to publish it to Mac AppStore.

**Comment by ameshkov** [comment]
@TheHasagi reopen the issue plz. Maybe in the future, there will be a way to have universal builds with Electron.

(or maybe we ditch Electron earlier).

**Comment by onodera-punpun** [comment]
>In the future, we will release a package that allows you to "merge" your arm64 and x64 apps into a single universal binary, but it's worth noting that this binary would be huge and probably isn't ideal for shipping to user

https://www.electronjs.org/blog/apple-silicon

Would it not just be easier to provide two separate versions of Aduard?

**Comment by RobbieTT** [comment]
Grateful for this being opened again since the plan to include it in the v1.9 milestone did not work out.

As Electron are proposing that packages are updated to both x64 and arm64 architectures is this split going to be adopted by Adguard or will Adguard be slowly depreciated with the Intel Mac retirement plan?

**Comment by ameshkov** [comment]
> Would it not just be easier to provide two separate versions of Aduard?

In this case, we'll have to publish two different apps to Mac App Store. We'd prefer to avoid this.

> Grateful for this being opened again since the plan to include it in the v1.9 milestone did not work out.

Well, there was some misunderstanding I guess. We have not planned to have a universal build in v1.9, just to make it work with Rosetta.

**Comment by RobbieTT** [comment]
Thanks and the current version works ok with Rosetta, in my experience at least.  Just like all the other non-native apps it just adds additional overheads that are best avoided.

Anyway, looking forward to native Apple Silicon projected build timeline.  I share your view that having 2 versions in the MAS is not desirable but I am in no position to judge the technical arguments for or against that approach.

Good luck!

**Comment by tkreind** [comment]
Would be greatly appreciated, thank you for the great extension!

**Comment by gnattu** [comment]
Electron-builder has merged universal support: https://github.com/electron-userland/electron-builder/pull/5481

Looks like this repo is using electron-packager, which has less feature.

**Comment by ameshkov** [comment]
Awesome, we'll see if we can use electron-builder.

@Mizzick is there any particular reason for not using it?

**Comment by Mizzick** [comment]
I guess we can try to switch to electron-builder.

**Comment by dlei** [comment]
Thanks for the great work! I'm curious if we have eta on releasing universal build? appreciated!

**Comment by ameshkov** [comment]
We'll check what can be done about it after v1.9 release.

**Comment by FutureMathematician** [comment]
> We'll check what can be done about it after v1.9 release.

Great!

**Comment by Macqael** [comment]
any (estimate) time for 1.10 ? (just to avoid another (any new on this ?) later from other user.

**Comment by ameshkov** [comment]
Tbh, initially, I thought we would at least have the first beta by now, but v1.9 takes much more time than anticipated. I can only say that this task will be the first we'll work on after v1.9 release.

**Comment by Macqael** [comment]
okay thanks for the answer if you want beta test I have a Mac mini with apple M1 so I can test (I will follow this thread)

**Comment by hadig** [comment]
Is it possible to release the iOS/iPadOS version for the Apple Silicon people? 
This would be a true native solution without electron at all

**Comment by s-valent** [comment]
> Is it possible to release the iOS/iPadOS version for the Apple Silicon people?
> This would be a true native solution without electron at all

@hadig True, fully native app would be really nice. However, I'm now currently using iOS AdGuard on my M1 mac and there are several issues with it:
1. Every time you reboot the mac you must launch the app (then you can close it).
2. It always writes that the filters cannot be updated.

(of course, as it's not officially supported, absolutely no complains here).

The point of my message is even if they switch to distributing iOS app on mac, it will need to be fixed in some aspects. + it will not be executable on Intel macs. Much better solution to native app would be not to use Electron, changing the interface of the app to SwiftUI.

**Comment by Macqael** [comment]
> > Is it possible to release the iOS/iPadOS version for the Apple Silicon people?
> > This would be a true native solution without electron at all
> 
> @hadig True, fully native app would be really nice. However, I'm now currently using iOS AdGuard on my M1 mac and there are several issues with it:
> 
> 1. Every time you reboot the mac you must launch the app (then you can close it).
> 2. It always writes that the filters cannot be updated.
> 
> (of course, as it's not officially supported, absolutely no complains here).
> 
> The point of my message is even if they switch to distributing iOS app on mac, it will need to be fixed in some aspects. + it will not be executable on Intel macs. Much better solution to native app would be not to use Electron, changing the interface of the app to SwiftUI. Although I don't really know how much of the app is using Electron, it can be too difficult to do.

I'm actually too for a Ui by Apple as customer but as Developper I actually know it's very annoying to maintain software with different Ui tech because it increase the update time delay two or tree fold (it depend of the update of course).
So I prefer an Electron app (I only want they compile it for M1 Ship).

**Comment by ameshkov** [comment]
There're some major under-the-hood differences between Mac and iOS. For instance, on Mac, we also maintain the "advanced blocking" and Safari toolbar icon extensions which simply won't work on iOS. At the same time, implementing DNS filtering the way it's done on iOS is pretty problematic on Macs (we're using a different way in the upcoming AG for Mac premium version update).

And finally, the main problem - M1 macs are still a tiny minority so we'll have to maintain three versions - normal iOS, iOS-ish M1 Mac, and Electron one. 

Maybe in the future when old Macs die out, this would be an option, but we are years from that moment.

**Comment by Macqael** [comment]
> There're some major under-the-hood differences between Mac and iOS. For instance, on Mac, we also maintain the "advanced blocking" and Safari toolbar icon extensions which simply won't work on iOS. At the same time, implementing DNS filtering the way it's done on iOS is pretty problematic on Macs (we're using a different way in the upcoming AG for Mac premium version update).
> 
> And finally, the main problem - M1 macs are still a tiny minority so we'll have to maintain three versions - normal iOS, iOS-ish M1 Mac, and Electron one.
> 
> Maybe in the future when old Macs die out, this would be an option, but we are years from that moment.

Actually it seems Apple want to move Mac from intel to their M ship within two year (security update only after that).

Having a Apple app who work on all at one it's a good idea BUT I'm not fan, a iOS version and a Mac version.
Using a the iOS version on Mac is not recommended by apple because iOS app on mac is only for those services who don't propose Mac version.

But if we forget iOS version on MAC M1, the better is compile Electron for M1 ship, and if possible change some settings of electron to optimise battery consumption/ cpu call and call it a day.
Because if you have more version to maintain it will only slow down update on all version of AdGuard.

**Comment by sweetppro** [comment]
why not just build a native macOS app and dump electron...
you can most likely reuse a lot of code from the iOS version

**Comment by Macqael** [comment]
if it's so simple to reuse this code why don't you do it and push it ?
IOS and MAC don't work at all in the same way, IOS have more Api to control the network (Mac haven't have them for now), since M1 is the first step to Import these more advanced security to Mac one app for (iOS (iPhone iPad) and MAC) is impossible for now.

**Comment by sweetppro** [comment]
> if it's so simple to reuse this code why don't you do it and push it ?
I never said it would be easy,
In fact It will be most likely a lot of work,
but native is definitely preferable to an electron based app and is more future proofed.

**Comment by ameshkov** [comment]
Well, we don't argue with that and eventually, we'd like to do that.

However, we're still developing 1.x, and we'd better use what we have.

**Comment by Macqael** [comment]
> > if it's so simple to reuse this code why don't you do it and push it ?
> 
> I never said it would be easy,
> In fact It will be most likely a lot of work,
> but native is definitely preferable to an electron based app and is more future proofed.

I totally agree, but the problem is the network stack of the app, since Mac use an old way to work, and iOS have a more secure way to do thing, it's actually not really possible to have one app for all, but I agree, switch from electron to swift can be cool, but I will increase update time, If they don't have a developer who work only on the swift app.

Mac app (I don't talk about this one but AdGuard for Mac), don't have all feature Windows one have (I think for exemple to V2 of malware API planned for later, I understand they are under staffed), so for now, electron help them to accelerate update so Swift ui universal app must wait they have finished to update all app to the same feature enabled.

Like I already said M1 accelerate our Mac and increase battery life so have an electron compiled for M1 will already be good don't be too greedy to ask a total rewrite of the UI who will take too much time.

**Comment by sgulls** [comment]
I understand that Universal is the preferred method but for those of us who are ok with bugs, being able to at least build a native binary from git would be appreciated in the short term by updating electron versions and other dependencies to ones that are arm64 compatible.

I'll try playing around with the code and will submit PRs if I can but I'm not quite proficient as many of you

**Comment by sgulls** [comment]
managed to get a thin binary app bundle going [in my fork](https://github.com/sgulls/AdGuardForSafari/tree/arm64)

**Comment by Macqael** [comment]
> managed to get a thin binary app bundle going [in my fork](https://github.com/sgulls/AdGuardForSafari/tree/arm64)

it means your fork is universal or apple silicon (arm64) ready (even with bug) ?

**Comment by sgulls** [comment]
The M1 universal [branch](https://github.com/AdguardTeam/AdGuardForSafari/tree/feature/m1_build) is working on my machine. 

Just change your Apple certificate dev info in `Scripts/ExportOptions.plist`, `AdGuard/AdGuard/Info.plist`, `AdGuard/Config.xcconfig`, and `AdGuard/AdGuard.xcodeproj/project.pbxproj`

**Comment by sgulls** [comment]
jk the universal part isn't working. added a few minor [fixes](https://github.com/AdguardTeam/AdGuardForSafari/pull/575) but it seems that the author of electron-universal [intended](https://github.com/electron/universal/issues/10#issuecomment-743728523) for binaries to be unsigned though the build process needs codesign in order to add entitlements.

it seems the build system needs to be overhauled in order to accommodate universal builds.

**Comment by sgulls** [comment]
Given the fact that universal binaries for electron are literally double the size, the lack of universal (fat) support so far is really only a Mac App Store distribution issue. Personally, I'd prefer the standalone bundle be only 300 MB instead of 600 MB so separate builds shouldn't be an issue. 

The only possible feature I can think of that separate builds for standalone release would necessitate is automatically installing or at least alerting the user to install arm64 over x64 if the program detects it suddenly is running translated with rosetta: `sysctl -in sysctl.proc_translated` (i.e. a user restored a x64 time machine backup onto a new arm Mac)

**Comment by Macqael** [comment]
any news on the release date of universal version (app silicon ready) ?

**Comment by tvinzz** [comment]
please check the universal build https://uploads.adguard.com/myMoon3122c.zip
unfortunately, there is a problem with Assistant for the moment, but it will be fixed in upcoming beta

**Comment by Oleg-Chashko** [comment]
Thanks for the universal build, it works stably until I found any problems other than those described above.

**Comment by TheHasagi** [comment]
@mcg The new [beta](https://github.com/AdguardTeam/AdGuardForSafari/releases/tag/v1.9.14-beta) is released with a universal build.

**Comment by Macqael** [comment]
> @mcg The new [beta](https://github.com/AdguardTeam/AdGuardForSafari/releases/tag/v1.9.14-beta) is released with a universal build.

Why closing it with beta ?! And not stable ??

**Comment by TheHasagi** [comment]
@Macqael 

Why not? The issue has already been resolved. The release is coming soon, but if you want to check out the latest features, try the beta version.

**Comment by tvinzz** [comment]
The issue was closed automatically by bot after new beta released. Please feel free to reopen it.

**Comment by mcg** [comment]
@TheHasagi Ah good to know. I expected to see a PR merged somewhere for that build. Am I missing that?

**Comment by ameshkov** [comment]
@mcg PRs take place in our internal self-hosted repo, it's already merged to the master branch.

**Comment by mcg** [comment]
@ameshkov Oh! Thanks for clarifying.

**Comment by RobbieTT** [comment]
I had a few issues getting this to run on Intel machines.  The app installed ok but on selecting settings I was greeted with a settings page without any content beyond the top line.  At this point the CPU load increased considerably.  After a few uninstalls / reinstalls, plus a number of annoying 'cannot install because' and 'cannot stop the app because' the app was showing as running but not shown in the Force Quit window, it was eventually coaxed into life after a few reboots.

No issues on my M1 Macs though.

**Comment by ameshkov** [comment]
> I had a few issues getting this to run on Intel machines. The app installed ok but on selecting settings I was greeted with a settings page without any content beyond the top line. At this point the CPU load increased considerably. After a few uninstalls / reinstalls, plus a number of annoying 'cannot install because' and 'cannot stop the app because' the app was showing as running but not shown in the Force Quit window, it was eventually coaxed into life after a few reboots.

Is it easy to reproduce? If it is, please open a new bug about that.

**Comment by Macqael** [comment]
@ameshkov Any estimation period of the update of the "store" version ? (just to know if i wait the release period, of if on my new mac (who don't have any third party app) i install this one as the first.

**Comment by ameshkov** [comment]
@Macqael a couple of weeks more. M1 support is not the only thing we're adding in this version. There will be at least one more beta version before pushing this update to store.

**Comment by Macqael** [comment]
> @Macqael a couple of weeks more. M1 support is not the only thing we're adding in this version. There will be at least one more beta version before pushing this update to store.

thanks take your time :), prefer wait, than having a buggy one 👍
